submissions/TaniaBondarenko/frogger_game/app.js

```diff
@@ -0,0 +1,120 @@
+const borders = {
+    rightBorder: 400,
+    leftBorder: 5,
+    upBorder: 410,
+    downBorder: 70
+};
+
+const someGap = 10;
+
+class GameSprite {
+    constructor (x, y,sprite)  {
+        this.sprite = sprite;
+        this.x = x;
+        this.y = y;
+    }
+
+    render () {
+        ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+    };
+}
+
+class Enemy extends GameSprite {
+    constructor(x, y, sprite = "images/enemy-bug.png", speed = "75") {
+        super(x, y, sprite);
+        this.speed = speed;
+    }
+ 
+    update(dt) {
+        this.x += this.speed * dt;
+        const addSpeed = 40;
+        const pointHideEnemy = 500;
+        if (this.x >= pointHideEnemy) {
+            this.x = -someGap;
+            this.speed = Math.floor(addSpeed + Math.random() * 100);
+        }
+        this.checkCollisions();
+    };
+
+    checkCollisions() {
+        allEnemies.forEach(function (enemy) {
+            const collision_distance = 50;
+            let diffX = Math.abs(enemy.x - player.x);
+            let diffY = Math.abs(enemy.y - player.y);
+            let distance = Math.sqrt(Math.pow(diffX, 2) + Math.pow(diffY, 2));
+            if (distance <= collision_distance) {
+                alert(`Oops... \uD83D\uDE3F Try Again`);
+                player.initialCoordinates();
+            } else if (player.y === borders.downBorder) {
+                player.y = someGap;
+                setTimeout(function () { alert(`\uD83D\uDE38 Congratulations! You won! \uD83D\uDE38 `); }, 100);
+                setTimeout(function () { player.initialCoordinates(); }, 200);
+            };
+        });
+    };
+};
```

Comment:
Issue 1. As this method is a part of the `Enemy` class it gets called for every enemy and yet runs through all enemies. Having 3 enemies total number of calls per every game tick is 3^3 = 9. Imagine we have 100 enemies. Then every game tick will lead to 100^100 = 10 000 calls. This will affect the performance.
Every enemy may check itself only against a player.

Issue 2. Class refers to a global variables `player` and `allEnemies`.
https://github.com/kottans/frontend-2021-homeworks/pull/299#discussion_r548045551

Reply:
1. Is it help if I replace `collision` method to `Player` class and apply `some` instead of `forEach` for `enemies`?
2. Even in class `Player` `this.player` refers to my global const player.

Reply:
1. If you move collision check (please make sure method names start with verbs) to `Player` then it will make sense to iterate through all enemies to conduct checks. It just doesn't make sense within `Enemy` class to run though all of them.

2. Cannot see `this.player` in `Player` class. And it would make any sense as `this` is already an instance of class `this` is used in.
What do you mean?

Reply:
1. Moved `checkCollision` to Player Class.  Invoked `this.checkCollision` in Player update method.
2.  if I leave it as below. player.x is referred to  - player (const player: Player).x ((property) Player.x: any). Is it what I need?
checkCollisions() {
        allEnemies.some(function (enemy) {
            const collision_distance = 50;
            let diffX = Math.abs(enemy.x - player.x);
            let diffY = Math.abs(enemy.y - player.y);
3. As `checkCollision` is a part of `class Player`, so I supposed that `this` should refer to  this player, not to global const player. But if I replace player.x by this.x, `this`  become undefined.